File: src_python/hello_rust/python_impl/d6.py
sample_input1 = sample_input2 = """Time:      7  15   30
Distance:  9  40  200""".splitlines()


def solve1(input: list[str]) -> str:
    times, dists = input
    times = list(map(int, times.split()[1:]))
    dists = list(map(int, dists.split()[1:]))

    ways_to_beat = 1
    for target_time, target_dist in zip(times, dists):
        speeds = []
        for charged_speed in range(1, target_time):
            dist = (target_time - charged_speed) * charged_speed
            if dist > target_dist:
                speeds.append(charged_speed)
        ways_to_beat *= len(speeds)

    return str(ways_to_beat)


def solve2(input: list[str]) -> str:
    target_time, target_dist = input
    target_time = int(target_time.replace("Time: ", "").replace(" ", ""))
    target_dist = int(target_dist.replace("Distance: ", "").replace(" ", ""))

    # Could to binary/ternary search but this is fast enough
    ways_to_beat = 0
    for charged_speed in range(1, target_time):
        dist = (target_time - charged_speed) * charged_speed
        if dist > target_dist:
            ways_to_beat += 1

    return str(ways_to_beat)


full_input1 = full_input2 = """Time:        38     67     76     73
Distance:   234   1027   1157   1236""".splitlines()

File: src/rust_impl/d6.rs
pub fn solve1(input: Vec<String>) -> String {
    let times = input.get(0).unwrap().replace("Time:", "");
    let dists = input.get(1).unwrap().replace("Distance:", "");
    let times = times
        .split_ascii_whitespace()
        .map(|n| n.parse::<i64>().unwrap())
        .collect::<Vec<_>>();

    let dists = dists
        .split_ascii_whitespace()
        .map(|n| n.parse::<i64>().unwrap())
        .collect::<Vec<_>>();

    let mut solutions = 1;
    for (dist, time) in std::iter::zip(dists, times) {
        let mut this_race_sol = 0;
        for charge_time in 1..time {
            let traversed = charge_time * (time - charge_time);
            if traversed > dist {
                this_race_sol += 1
            }
        }
        solutions *= this_race_sol;
    }
    solutions.to_string()
}

pub fn solve2(input: Vec<String>) -> String {
    let times = input.get(0).unwrap().replace("Time:", "");
    let dists = input.get(1).unwrap().replace("Distance:", "");
    let time = times.replace(" ", "").parse::<i64>().unwrap();
    let dist = dists.replace(" ", "").parse::<i64>().unwrap();

    let mut solutions = 1;
    let mut this_race_sol = 0;
    for charge_time in 1..time {
        let traversed = charge_time * (time - charge_time);
        if traversed > dist {
            this_race_sol += 1
        }
    }
    solutions *= this_race_sol;
    solutions.to_string()
}

pub fn sample_input1() -> Vec<String> {
    vec![
        "Time:      7  15   30".to_string(),
        "Distance:  9  40  200".to_string(),
    ]
}

pub fn sample_input2() -> Vec<String> {
    sample_input1()
}
pub fn full_input1() -> Vec<String> {
    full_input2()
}

pub fn full_input2() -> Vec<String> {
    vec![
        "Time:        38     67     76     73".to_string(),
        "Distance:   234   1027   1157   1236".to_string(),
    ]
}
